### ISSUE TITLE
Argument of type 'NoneType' is not iterable while running on Docker/airflow

### ISSUE THREAD

**Issue by tannu14** [yt-dlp/yt-dlp#8816]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

Extracting channel metadata like this:

```py
channel_url = f'https://www.youtube.com/channel/{channel_id}'
with YoutubeDL() as ydl:
    result = ydl.extract_info(channel_url, download=False)
    logging.info("Youtube post extraction data %s", result)
```

======
In local this code is working fine able to get the results. As soon as I started running through docker/airflow. I started getting this error.
It is not even going on logging.info and throwing this error:

```
[2023-12-20, 20:26:54 UTC] {logging_mixin.py:137} INFO - offset_val**: 2023-12-08T22:51:16.332110Z
[2023-12-20, 20:26:54 UTC] {youtube.py:285} INFO - youtube_creator_offset_key: youtube_data_ingestion_28020_offset, offset_val in cache: 2023-12-08T22:51:16.332110Z, published_after: 2023-12-08T22:51:16.332110Z
[2023-12-20, 20:26:54 UTC] {youtube.py:815} ERROR - argument of type 'NoneType' is not iterable
Traceback (most recent call last):
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/youtube.py", line 766, in process_channel_videos
    with YoutubeDL(ydl_opts) as ydl:
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 681, in __init__
    self.print_debug_header()
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 3922, in print_debug_header
    write_string(f'[Invalid date] {encoding_str}\n', encoding=None)
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/utils/_utils.py", line 1429, in write_string
    if 'b' in getattr(out, 'mode', ''):
TypeError: argument of type 'NoneType' is not iterable
[2023-12-20, 20:26:55 UTC] {youtube.py:298} ERROR - ('Error while scraping youtube posts: %s', 'UCAcr6uli4eZhYKyMrKPB87A')
Traceback (most recent call last):
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/youtube.py", line 766, in process_channel_videos
    with YoutubeDL(ydl_opts) as ydl:
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 681, in __init__
    self.print_debug_header()
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 3922, in print_debug_header
    write_string(f'[Invalid date] {encoding_str}\n', encoding=None)
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/utils/_utils.py", line 1429, in write_string
    if 'b' in getattr(out, 'mode', ''):
TypeError: argument of type 'NoneType' is not iterable
```

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
*** Reading local file: /usr/local/airflow/logs/dag_id=youtube_data_ingestion_dag/run_id=manual__2023-12-20T20:21:40.071276+00:00/task_id=fetch_and_process_social_content/map_index=0/attempt=2.log
[2023-12-20, 20:26:54 UTC] {taskinstance.py:1165} INFO - Dependencies all met for <TaskInstance: youtube_data_ingestion_dag.fetch_and_process_social_content manual__2023-12-20T20:21:40.071276+00:00 map_index=0 [queued]>
[2023-12-20, 20:26:54 UTC] {taskinstance.py:1165} INFO - Dependencies all met for <TaskInstance: youtube_data_ingestion_dag.fetch_and_process_social_content manual__2023-12-20T20:21:40.071276+00:00 map_index=0 [queued]>
[2023-12-20, 20:26:54 UTC] {taskinstance.py:1362} INFO - 
--------------------------------------------------------------------------------
[2023-12-20, 20:26:54 UTC] {taskinstance.py:1363} INFO - Starting attempt 2 of 2
[2023-12-20, 20:26:54 UTC] {taskinstance.py:1364} INFO - 
--------------------------------------------------------------------------------
[2023-12-20, 20:26:54 UTC] {taskinstance.py:1383} INFO - Executing <Mapped(_PythonDecoratedOperator): fetch_and_process_social_content> on 2023-12-20 20:21:40.071276+00:00
[2023-12-20, 20:26:54 UTC] {standard_task_runner.py:55} INFO - Started process 1860 to run task
[2023-12-20, 20:26:54 UTC] {standard_task_runner.py:82} INFO - Running: ['airflow', 'tasks', 'run', 'youtube_data_ingestion_dag', 'fetch_and_process_social_content', 'manual__2023-12-20T20:21:40.071276+00:00', '--job-id', '140', '--raw', '--subdir', 'DAGS_FOLDER/data_ingestion_dag.py', '--cfg-path', '/tmp/tmpbt7fbvi3', '--map-index', '0']
[2023-12-20, 20:26:54 UTC] {standard_task_runner.py:83} INFO - Job 140: Subtask fetch_and_process_social_content
[2023-12-20, 20:26:54 UTC] {task_command.py:376} INFO - Running <TaskInstance: youtube_data_ingestion_dag.fetch_and_process_social_content manual__2023-12-20T20:21:40.071276+00:00 map_index=0 [running]> on host 7d8ef6825cf3
[2023-12-20, 20:26:54 UTC] {taskinstance.py:1590} INFO - Exporting the following env vars:
AIRFLOW_CTX_DAG_OWNER=airflow
AIRFLOW_CTX_DAG_ID=youtube_data_ingestion_dag
AIRFLOW_CTX_TASK_ID=fetch_and_process_social_content
AIRFLOW_CTX_EXECUTION_DATE=2023-12-20T20:21:40.071276+00:00
AIRFLOW_CTX_TRY_NUMBER=2
AIRFLOW_CTX_DAG_RUN_ID=manual__2023-12-20T20:21:40.071276+00:00
[2023-12-20, 20:26:54 UTC] {youtube.py:272} INFO - dag_id: youtube_data_ingestion_dag, Starting YouTube posts ingestion for the viral_nation_id: 28020
[2023-12-20, 20:26:54 UTC] {logging_mixin.py:137} INFO - offset_val**: 2023-12-08T22:51:16.332110Z
[2023-12-20, 20:26:54 UTC] {youtube.py:285} INFO - youtube_creator_offset_key: youtube_data_ingestion_28020_offset, offset_val in cache: 2023-12-08T22:51:16.332110Z, published_after: 2023-12-08T22:51:16.332110Z
[2023-12-20, 20:26:54 UTC] {youtube.py:815} ERROR - argument of type 'NoneType' is not iterable
Traceback (most recent call last):
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/youtube.py", line 766, in process_channel_videos
    with YoutubeDL(ydl_opts) as ydl:
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 681, in __init__
    self.print_debug_header()
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 3922, in print_debug_header
    write_string(f'[Invalid date] {encoding_str}\n', encoding=None)
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/utils/_utils.py", line 1429, in write_string
    if 'b' in getattr(out, 'mode', ''):
TypeError: argument of type 'NoneType' is not iterable
[2023-12-20, 20:26:55 UTC] {youtube.py:298} ERROR - ('Error while scraping youtube posts: %s', 'UCAcr6uli4eZhYKyMrKPB87A')
Traceback (most recent call last):
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/youtube.py", line 766, in process_channel_videos
    with YoutubeDL(ydl_opts) as ydl:
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 681, in __init__
    self.print_debug_header()
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 3922, in print_debug_header
    write_string(f'[Invalid date] {encoding_str}\n', encoding=None)
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/utils/_utils.py", line 1429, in write_string
    if 'b' in getattr(out, 'mode', ''):
TypeError: argument of type 'NoneType' is not iterable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/youtube.py", line 292, in youtube_fetch_social_content_using_scraping
    process_channel_videos(viral_nation_id, 'UCAcr6uli4eZhYKyMrKPB87A', published_after, user_to_process)
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/youtube.py", line 818, in process_channel_videos
    raise Exception(
Exception: ('Error while scraping youtube posts: %s', 'UCAcr6uli4eZhYKyMrKPB87A')
[2023-12-20, 20:26:55 UTC] {taskinstance.py:1851} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/youtube.py", line 766, in process_channel_videos
    with YoutubeDL(ydl_opts) as ydl:
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 681, in __init__
    self.print_debug_header()
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 3922, in print_debug_header
    write_string(f'[Invalid date] {encoding_str}\n', encoding=None)
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/utils/_utils.py", line 1429, in write_string
    if 'b' in getattr(out, 'mode', ''):
TypeError: argument of type 'NoneType' is not iterable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/youtube.py", line 292, in youtube_fetch_social_content_using_scraping
    process_channel_videos(viral_nation_id, 'UCAcr6uli4eZhYKyMrKPB87A', published_after, user_to_process)
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/youtube.py", line 818, in process_channel_videos
    raise Exception(
Exception: ('Error while scraping youtube posts: %s', 'UCAcr6uli4eZhYKyMrKPB87A')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/airflow/decorators/base.py", line 188, in execute
    return_value = super().execute(context)
  File "/usr/local/lib/python3.9/site-packages/airflow/operators/python.py", line 175, in execute
    return_value = self.execute_callable()
  File "/usr/local/lib/python3.9/site-packages/airflow/operators/python.py", line 193, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/usr/local/airflow/dags/data_ingestion_dag.py", line 75, in fetch_and_process_social_content
    fetch_and_process_social_content_data(dag, dag_id, configs,
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/processor/fetch_social_content_processor.py", line 50, in fetch_and_process_social_content_data
    youtube_fetch_social_content_using_scraping(dag, dag_id, configs, job_type_cfg, influencer)
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/youtube.py", line 299, in youtube_fetch_social_content_using_scraping
    raise Exception(
Exception: Error while fetching video posts jsons
[2023-12-20, 20:26:55 UTC] {taskinstance.py:1401} INFO - Marking task as FAILED. dag_id=youtube_data_ingestion_dag, task_id=fetch_and_process_social_content, map_index=0, execution_date=20231220T202140, start_date=20231220T202654, end_date=20231220T202655
[2023-12-20, 20:26:55 UTC] {standard_task_runner.py:100} ERROR - Failed to execute job 140 for task fetch_and_process_social_content (Error while fetching video posts jsons; 1860)
[2023-12-20, 20:26:55 UTC] {local_task_job.py:159} INFO - Task exited with return code 1
[2023-12-20, 20:26:55 UTC] {taskinstance.py:2623} INFO - 0 downstream tasks scheduled from follow-on schedule check
```

**Comment by tannu14** [comment]
I went through this also: https://github.com/ytdl-org/youtube-dl/issues/20737
checked that my channel/video is not blocked.

**Comment by bashonly** [comment]
Can you get the yt-dlp output? Instead of the output from your script

Are you running this on a data center IP?

**Comment by tannu14** [comment]
This is the yt-dlp output:

2023-12-21, 01:38:22 UTC] {youtube.py:307} ERROR - argument of type 'NoneType' is not iterable
Traceback (most recent call last):
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/youtube.py", line 301, in youtube_fetch_social_content_using_scraping
    scrap_channel(viral_nation_id, 'UCAcr6uli4eZhYKyMrKPB87A', published_after, user_to_process)
  File "/usr/local/airflow/dags/tasks/brandsafety/openreport/youtube.py", line 847, in scrap_channel
    info = ydl.extract_info(channel_url, download=False)
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 1556, in extract_info
    return self.__extract_info(url, self.get_info_extractor(key), download, extra_info, process)
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 1567, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 1702, in __extract_info
    ie_result = ie.extract(url)
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 713, in extract
    self.to_screen('Extracting URL: %s' % (
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 1136, in to_screen
    self._downloader.to_screen(f'[Invalid date] {msg}', *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 898, in to_screen
    self._write_string(
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 879, in _write_string
    write_string(message, out=out, encoding=self.params.get('encoding'))
  File "/usr/local/lib/python3.9/site-packages/yt_dlp/utils/_utils.py", line 1429, in write_string
    if 'b' in getattr(out, 'mode', ''):
TypeError: argument of type 'NoneType' is not iterable
===========
I am running is on :Airflow Webserver: http://localhost:8080

**Comment by Grub4K** [comment]
This still seems like either a problem with your environment or the runner. What are you passing as `YoutubeDL` options?

Currently, we assume if `out.mode` exists, it has to be a `str` (as defined by the `FileIO` interface.
Perhaps we should check for `isinstance(out, (RawIOBase, BufferedIOBase))` instead? That defeats duck typing however.
Another simple fix would be this:
```diff
diff --git a/yt_dlp/utils/_utils.py b/yt_dlp/utils/_utils.py
index f1d7cead6..b3f4fc1a7 100644
--- a/yt_dlp/utils/_utils.py
+++ b/yt_dlp/utils/_utils.py
@@ -1424,7 +1424,7 @@ def write_string(s, out=None, encoding=None):
         s = re.sub(r'([\r\n]+)', r' \1', s)

     enc, buffer = None, out
-    if 'b' in getattr(out, 'mode', ''):
+    if 'b' in (getattr(out, 'mode') or ''):
         enc = encoding or preferredencoding()
     elif hasattr(out, 'buffer'):
         buffer = out.buffer
```
This could give other problems though as the interface is already not followed correctly. Maybe its better to `try except`?

**Comment by pukkandan** [comment]
> `File "/usr/local/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 1136, in to_screen`
> `    self._[downloader.to](http://downloader.to)_screen(f'[Invalid date] {msg}', *args, **kwargs)`

This is not code that exists in yt-dlp. Are you using a modified version?

> What are you passing as `YoutubeDL` options?

Nothing according to OP

```py
channel_url = f'https://www.youtube.com/channel/{channel_id}'
with YoutubeDL() as ydl:
    result = ydl.extract_info(channel_url, download=False)
    logging.info("Youtube post extraction data %s", result)
```


@Grub4K's patch is reasonable to merge, but there's something weird going on here. What does `ydl._out_files` give you?

**Comment by pukkandan** [comment]
> ```python
>   File "/usr/local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 3922, in print_debug_header
>     write_string(f'[Invalid date] {encoding_str}\n', encoding=None)
> ```
> 
> This is not code that exists in yt-dlp. Are you using a modified version?


This seems to be what the traceback is trying to refer to: https://github.com/yt-dlp/yt-dlp/blob/298230e550886b746c266724dd701d842ca2696e/yt_dlp/YoutubeDL.py#L3939

**Comment by Grub4K** [comment]
> ...`mode` is only defined for `FileIO` instances. The `sys.std...` streams are specified to be "file-like" which might not be the same thing.

This can be proven wrong as per the [Python 3 glossary](<https://docs.python.org/3/glossary.html#term-file-object>) which states:
> There are actually three categories of file[-like] objects: raw [binary files](https://docs.python.org/3/glossary.html#term-binary-file), buffered [binary files](https://docs.python.org/3/glossary.html#term-binary-file) and [text files](https://docs.python.org/3/glossary.html#term-text-file). Their interfaces are defined in the [io](https://docs.python.org/3/library/io.html#module-io) module

So, per definition, having an object there that is not `None` but does not duck type as any of these should be incorrect. The question is *why* though, where does that object come from? If it doesn't follow specifications here, it could give strange values for others as well, like `TextIOBase.buffer` later on in the code. Finding the root cause is more pressing, so we can gauge if it's our problem or the consumer's problem.

> This is supposing, in the last line, that it really is correct to override the encoding of the stream with the encoding passed into `write_string()`.

There are two reasons to manually encode it here:
1. We want to write to the stream without raising errors. This can be done by manually using the same encoding and using `encode(..., 'ignore')`
2. While wanting to obey the encoding (hence trying to get encoding and falling back to a sensible `preferredencoding()`), there are cases where we want to write to the stream uncompromised. This case is relevant when forcing encoding (`--encoding`).

Perhaps `write_string` could have some additional documentation with reasoning behind the things it does, since they are not immediately obvious.

**Comment by tannu14** [comment]
@Grub4K @dirkf - I am not sure if I am able to follow the conversation. Is there any thing incorrect with my usage? Is there anything I can do to unblock myself?
This code is working on my local machine, the moment I put this on airflow dag (running on my local docker container), it gives me this error.

**Comment by Grub4K** [comment]
It seems as if airflow replaces `sys.stdout` and `sys.stderr` with file-like objects that don't adhere to the convention that `FileIO.mode` should always be of type `str`. The main cases are these:
- Having `FileIO.mode` be `None` is common practice (yt-dlp should apply any of the workarounds mentioned above)
- This is the default for airflows `sys.stderr` or their replacement (an issue should be raised on their tracker)
- Your code is somehow creating the workflow incorrectly causing `sys.stderr` to have a `mode` of `None` (then your code has to be adapted)

If you just need an immediate workaround, you can try this in your code:
```py
# Work around sys.stdout wrapper having a mode of `None`
if getattr(sys.stdout, "mode", ...) is None:
    sys.stdout.mode = "b"
if getattr(sys.stderr, "mode", ...) is None:
    sys.stderr.mode = "b"
```
If the above causes a TypeError you can try `""` instead of `"b"`, or perhaps `"bw+"`, `"w+"`, `"bw"` or `"w"`.

**Comment by tannu14** [comment]
So these lines of  code:
https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/utils/_utils.py#L1426-L1432
Should I replsce it like this:
<img width="756" alt="image" src="https://github.com/yt-dlp/yt-dlp/assets/22826610/46318554-56f3-4338-a05b-f522dbd5ed82">

Does it look correct to you?
Also If I make changes in my local yt_dlp/utils/_utils.py, will it work?

**Comment by Grub4K** [comment]
no you should place these lines in your own code, before the `with YoutubeDL() as ydl:` part. alternatively apply [this](<https://github.com/yt-dlp/yt-dlp/issues/8816#issuecomment-1866187363>) if you know what you are doing and want to modify the installed python library

**Comment by dirkf** [comment]
Regarding case 1 [above](https://github.com/yt-dlp/yt-dlp/issues/8816#issuecomment-1866742916), the `write_string()` code is looking for any `Object.mode`, not just `FileIO.mode`, so the expected semantics might not apply. Eg, a custom subclass of `TextIOBase` might be given a nullable `mode` attribute; a duck-equivalent class similarly (it doesn't quack like a `FileIO` because it's a `TextIOBase` duck).

Although this is hardly "common practice", so case 2 applies, the _airflow_ authors could be forgiven for not realising the problem. Eg if their `open()` makes `mode` have a `None` default, they might have copied the value through to the stream object instead of omitting it or changing it to `'w'`. In case others have the same confusion, some small defences could applied even against case 2.

Unrelated, the current code uses `buffer` to identify an `out` that has an `encoding`. According to the doc, that is not reliable ("... not part of the `TextIOBase` API and may not exist in some implementations") and it would be better to test for `encoding` itself as in my diff above.

**Comment by Grub4K** [comment]
The codes ideas are not very obvious, which is primarily why I requested clarifying comments. That way the intent behind some of the decisions could be more easily translated into generically robust code.

> `write_string()` code is looking for any `Object.mode`, not just `FileIO.mode`

Hmm, since we are expecting a file object as input it should follow one of the interfaces, even if duck typed. Nothing however forces for example a duck typed `RawIOBase` impl to *not* have an additional `mode` so you are correct. If we take it to that point then most of `write_string` is incorrectly assuming interfaces behave the way we expect them to. You have pointed that out in your last paragraph as well.

> Although this is hardly "common practice", so case 2 applies, the *airflow* authors could be forgiven for not realising the problem.

The Python APIs seem to disagree with this sentiment:
```py
>>> open(0, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: open() argument 'mode' must be str, not None
>>> import io
>>> io.FileIO(0, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: FileIO() argument 'mode' must be str, not None
```

However I agree that this function has to be secured better against some of these cases.

The output rework did as much, additionally probing the stream only once and wrapping it its own helper class. I should really revive that PR and the related Logger rework to start providing more robust functionality regarding these

**Comment by tannu14** [comment]
> ce these lines in your own code, b

I tried as you suggested, throwing another error:

[2023-12-22, 06:04:27 UTC] {youtube.py:718} ERROR - can't set attribute
Traceback (most recent call last):
  File "/usr/local/airflow/dags/tasks/brandsafety/common_code/youtube.py", line 666, in process_channel_videos
    sys.stdout.mode = "b"
AttributeError: can't set attribute
[2023-12-22, 06:04:27 UTC] {youtube.py:539} ERROR - ('Error while scraping youtube posts: %s', 'UCAcr6uli4eZhYKyMrKPB87A')
Traceback (most recent call last):
  File "/usr/local/airflow/dags/tasks/brandsafety/common_code/youtube.py", line 666, in process_channel_videos
    sys.stdout.mode = "b"
AttributeError: can't set attribute

**Comment by dirkf** [comment]
Since you can't set an attribute on the system's `sys.stderr`, you could try wrapping it in an object of a custom class (which has an empty `mode` attribute instead of `None`) and setting `sys.stderr` to that object.

[This SO question](https://stackoverflow.com/questions/1443129/completely-wrap-an-object-in-python) gives examples.

**Comment by botsmaster** [comment]
👋 Hi guys!  I have a similar problem, but when calling the code:

```
  proxy = {"socks": "socks5://cTM00:zjBgtv@538.279.120.28:1163"}
    options = {
        'proxy' : proxy 
         
    }
    link = 'https://www.youtube.com/shorts/wjxflqVyKh4'
    with ytd.YoutubeDL(options) as ytdl:
        result = ytdl.extract_info(link, download=False)
        # with open('you2.txt', 'w', encoding='utf-8') as file:
        #     file.write(str(result))
        if 'formats' in result:
            for format_info in result['formats']:
                format_id = format_info.get('format_id')
                filesize = format_info.get('filesize')
       print(format_id , filesize )
```

I get all file sizes except 18 and 22. I have them marked as None. And the most interesting thing is that this started happening on one server after a certain number of requests, on the other everything is fine. Using a proxy did not help. Maybe someone knows why this is happening and how to solve the problem of getting the sizes of formats 18 and 22? I will be very grateful!!🙏

**Comment by pukkandan** [comment]
> What does `ydl._out_files` give you?

@tannu14

**Comment by pukkandan** [comment]
@botsmaster Your comment is not related to this in any way. Open a new issue with appropriate verbose log

**Comment by tannu14** [comment]
Hello @Grub4K ,
Is there any provision to push a fix on this?
Will I be able to use yt_dlp library for my usecase?

**Comment by Grub4K** [comment]
If you follow pukkandans request and show us what `ydl._out_files` is we can start debugging

**Comment by tannu14** [comment]
Hello @pukkandan @Grub4K 

As per my understanding, To specifically catch errors related to ydl._out_files, modified my except block to capture ExtractorError, which is a specific exception class that youtube_dl raises. Here’s how you I did it:
<img width="987" alt="image" src="https://github.com/yt-dlp/yt-dlp/assets/22826610/138d3a62-5551-4de7-a302-94162f89b86c">

but it is not capturing any ExtractorError exception.

Please let me knwo if I am missing naything overe here. Or is there any other way to capture ydl._out_files logs.

**Comment by tannu14** [comment]
I tried to capture these exceptions like this:
<img width="777" alt="image" src="https://github.com/yt-dlp/yt-dlp/assets/22826610/7aa4c8f3-2af6-4ca0-bbcd-8a7a02f1f114">

[2023-12-28, 16:47:27 UTC] {youtube.py:687} ERROR - TypeError: argument of type 'NoneType' is not iterable
[2023-12-28, 16:47:27 UTC] {python.py:177} INFO - Done. Returned value was: None

**Comment by bashonly** [comment]
we just need to see what your `ydl._out_files` are

all you need to do is something like this:

```py
with YoutubeDL(ydl_opts) as ydl:
    logging.error(ydl._out_files)
```

**Comment by tannu14** [comment]
[2023-12-28, 18:25:19 UTC] {youtube.py:678} ERROR - ydl._out_files: Namespace(out=<airflow.utils.log.logging_mixin.StreamLogWriter object at 0x7fae8ae115b0>, error=<airflow.utils.log.logging_mixin.StreamLogWriter object at 0x7fae8ae116a0>, screen=<airflow.utils.log.logging_mixin.StreamLogWriter object at 0x7fae8ae116a0>, console=None)

**Comment by Grub4K** [comment]
`airflow.utils.log.logging_mixin.StreamLogWriter` subclasses `typing.IO[str]`, which seems to be a viable way to create the text and binary writers the simple way.
```py
import typing

print(typing.IO[str]().mode)
print(typing.IO[bytes]().mode)
```
Given this short snippet prints `None` for both, I think fixing `write_string` against that value only would make the most sense.

However, since we cannot potentially know if the object will be binary or not, this could still cause errors down the line. I think this could only be handeled in a sane by output rework, given that it can check write calls and save it in the class instance. While this is possible for `write_string` to do as well, it will not be as pretty. I can provide an implementation for the more robust `write_string` rewrite if @pukkandan thinks it would make sense

**Comment by dirkf** [comment]
A more robust `write_string()` would be good and could work around this and similar issues; but, based on the earlier discussion, isn't this actually a bug that should be reported against the `typing` Python library module ("ABCs for working with IO")?

**Comment by pukkandan** [comment]
> Given this short snippet prints None for both, I think fixing write_string against that value only would make the most sense.

Imo, we should do https://github.com/yt-dlp/yt-dlp/issues/8816#issuecomment-1866187363 for now and do proper handling in output rework. Just make sure to note it down in your PR so we don't forget

**Comment by Grub4K** [comment]
I stand corrected by the CPython source code:
```py
class IO(Generic[AnyStr]):
    @property
    @abstractmethod
    def mode(self) -> str:
        pass
```
Theoretically it should be required to override the `mode` property to return a `str` since it is abstract. However, that doesn't seem to be the case in practice, so this is indeed an external issue. The ordering of the subclasses in `airflow` matters: Subclassing `IOBase, typing.IO[str]` does not throw a type error (`abstract`) using mypy, while subclassing `typing.IO[str], IOBase` does.

This makes me think that our code behaves as expected and the issue should be raised on the `airflow` repo. I will still commit the workaroud patch soon:tm:

**Comment by tannu14** [comment]
Thank you so much team for taking efforts on this issue. @Grub4K @pukkandan @dirkf please let me know once you push the changes. I will retest it from my end.

**Comment by tannu14** [comment]
Any update on this @Grub4K ?

**Comment by Grub4K** [comment]
I have opened https://github.com/apache/airflow/issues/36992 and the patch will make it into the [cleanup PR](<https://github.com/yt-dlp/yt-dlp/pull/8968>) (reminder for myself)
